fix: userCanPay without existing payment methods

josh-burton opened the pull request:
Fixes #305 

**Describe the changes proposed**<br>
Adds a new `existingPaymentMethodRequired` parameter to the `userCanPay` method, so the developer can determine if Apple/Google pay is available without requiring cards to be added

**Screenshots / Videos**<br>
If useful, include any screeshots or screencasts to better explain the change.

**Additional context**<br>
Add any other context about the change here.

josh-burton committed:
fix: #305 userCanPay can now determine if user can pay without requiring cards to be added

Adds a new existingPaymentMethodRequired parameter to the userCanPay method, so the developer can determine if Apple/Google pay is available without requiring cards to be added

josh-burton committed:
fix: #305 userCanPay can now determine if user can pay without requiring cards to be added

Adds a new existingPaymentMethodRequired parameter to the userCanPay method, so the developer can determine if Apple/Google pay is available without requiring cards to be added

JlUgia commented:
Hi @josh-burton, thank you for your contributions.
My preference would be to default to `false` on this new prop, in line with how both enabled providers operate today.
Also, how about we use a more straightforward prop name on the Flutter API surface? Both Apple and Google Pay have slight variations on how the API works, so I feel that something like `existingPaymentMethodAvailable` or similar would be more accurate, since we are looking to extract whether there's at least one payment method available for the configuration provided.

Thank you again.

josh-burton committed:
fix: renames existingPaymentMethodRequired to existingPaymentMethodAvailable, sets default values to false

josh-burton committed:
fix: renames existingPaymentMethodRequired to existingPaymentMethodAvailable, sets default values to false

josh-burton commented:
> Hi @josh-burton, thank you for your contributions. My preference would be to default to `false` on this new prop, in line with how both enabled providers operate today. Also, how about we use a more straightforward prop name on the Flutter API surface? Both Apple and Google Pay have slight variations on how the API works, so I feel that something like `existingPaymentMethodAvailable` or similar would be more accurate, since we are looking to extract whether there's at least one payment method available for the configuration provided.
> 
> Thank you again.

Thank you for the review. I have made those changes

JlUgia commented on pay_ios/ios/Classes/PaymentHandler.swift:
```diff
@@ -36,26 +36,31 @@ enum PaymentHandlerStatus {
     if let supportedNetworks = PaymentHandler.supportedNetworks(from: paymentConfiguration) {
-      return PKPaymentAuthorizationController.canMakePayments(usingNetworks: supportedNetworks)
+      if existingPaymentMethodAvailable {
+        return PKPaymentAuthorizationController.canMakePayments(usingNetworks: supportedNetworks)
```
If `supportedNetworks` is only used here, is there any value in creating the object beforehand, and returning false if no `supportedNetworks` are set?

JlUgia commented on pay_android/android/src/main/kotlin/io/flutter/plugins/pay_android/GooglePayHandler.kt:
```diff
@@ -138,9 +139,11 @@ class GooglePayHandler(private val activity: Activity) : PluginRegistry.Activity
             )
         )
+        // Inject existingPaymentMethodAvailable into the profile
+        paymentProfile.put("existingPaymentMethodAvailable", existingPaymentMethodAvailable)
```
I'm assuming we'll want to set this every time we build a payment profile. The `buildPaymentProfile` can be the right place to do so. Also, on Android, this requirement is enforced by the `existingPaymentMethodRequired` (see [docs](https://developers.google.com/pay/api/android/reference/request-objects#IsReadyToPayRequest)).

JlUgia commented on pay/lib/src/pay.dart:
```diff
@@ -47,10 +47,14 @@ class Pay {
     await throwIfProviderIsNotDefined(provider);
     if (supportedProviders[defaultTargetPlatform]!.contains(provider)) {
-      return _payPlatform.userCanPay(_configurations[provider]!);
+      return _payPlatform.userCanPay(_configurations[provider]!, existingPaymentMethodAvailable: existingPaymentMethodAvailable);
```
If this field is already part of the configuration, does it need to be also passed a method parameter here?

josh-burton replied:
We need this as a method parameter to allow some call sites only check whether Google Pay is supported, while others also need to check that the user has a card available (while using a single JSON config)

josh-burton committed:
fix: address payment method availability review feedback

- align the availability flag with existingPaymentMethodRequired
- map the requirement when building Android payment profiles
- avoid parsing Apple Pay networks for device-support-only checks
- improve availability documentation and update channel tests
